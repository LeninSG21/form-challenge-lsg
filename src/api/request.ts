import axios, { AxiosRequestConfig } from 'axios';

const errorParser = (statusCode: number, statusText: string) => {
  if (statusCode >= 500) {
    return '[500] Something went wrong. Try again';
  }
  if (statusCode >= 400) {
    if (statusCode === 400) {
      return '[400] Bad request';
    }
    if (statusCode === 401) {
      return '[401] Unauthorized';
    }
  }
  return `[${statusCode}]: ${statusText}`;
};

export const abortController = new AbortController();

type RequesOptions = {
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: unknown;
};

type RequestResponse<T> = {
  data: T;
  status: number;
};

const DEFAULT_REQUEST_OPTIONS: RequesOptions = {
  method: 'GET',
};

export async function request<T = undefined>(
  url: string,
  options?: RequesOptions,
): Promise<RequestResponse<T>> {
  const optionsWithDefault = { ...DEFAULT_REQUEST_OPTIONS, ...options };

  const instance = axios.create();
  const axiosOptions: AxiosRequestConfig = {
    url,
    method: optionsWithDefault.method,
    data: optionsWithDefault.body,
  };

  return instance
    .request<T>(axiosOptions)
    .then((response) => ({ data: response.data, status: response.status }))
    .catch((error) => {
      throw new Error(
        errorParser(error.response.status, error.response.statusText),
      );
    });
}
