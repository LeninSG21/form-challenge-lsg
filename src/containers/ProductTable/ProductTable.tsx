import { Table } from '../../components/Table';
import { useProducts } from '../../hooks/useProducts';
import { COLUMNS } from './columns.config';

export function ProductTable() {
  const { products, isLoading, isRejected, fetchProducts } = useProducts();

  if (isLoading) return <div>Loading data...</div>;
  if (isRejected) {
    return (
      <div>
        Something went wrong. Please, try again.{' '}
        <button type="button" onClick={fetchProducts}>
          Retry
        </button>
      </div>
    );
  }
  return <Table columns={COLUMNS} data={products} />;
}
