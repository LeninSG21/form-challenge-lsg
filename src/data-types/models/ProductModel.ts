export interface ProductModel {
  code: string;
  position: number;
  quantity: number;
  image: string;
  price: number;
  description: string;
}

export type ProductsModel = ProductModel[];
