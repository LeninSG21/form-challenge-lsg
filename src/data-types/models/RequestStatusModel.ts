export const REQUEST_STATUS = {
  idle: 'idle',
  pending: 'pending',
  resolved: 'resolved',
  rejected: 'rejected',
} as const;

export type RequestStatusType =
  typeof REQUEST_STATUS[keyof typeof REQUEST_STATUS];
