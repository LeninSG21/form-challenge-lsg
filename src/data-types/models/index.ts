export * from './ProductModel';
export * from './RequestStatusModel';
