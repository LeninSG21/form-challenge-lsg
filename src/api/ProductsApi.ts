import { ProductModel } from '../data-types/models';
import products from './products.json';

//* Needless to say, it is an utility to function to mock an error
const generateRandomeError = () => Math.ceil(Math.random() * 10) % 4 === 0;

export const ProductsApi = {
  getProducts() {
    return new Promise<ProductModel[]>((res, rej) => {
      setTimeout(() => {
        if (generateRandomeError()) rej(new Error('Internal Server Error'));
        res(products.products);
      }, 1000);
    });
  },
};
