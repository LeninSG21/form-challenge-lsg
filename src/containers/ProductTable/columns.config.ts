import { ExtendedColumn } from '../../components/Table';
import { ProductModel } from '../../data-types/models';

export const COLUMNS: ExtendedColumn<ProductModel>[] = [
  {
    Header: 'Code',
    accessor: (props) => props.code,
  },
  {
    Header: 'Position',
    accessor: (props) => props.position,
  },
  {
    Header: 'Quantity',
    accessor: (props) => props.quantity,
  },
  {
    Header: 'Price',
    accessor: (props) => props.price,
  },
  {
    Header: 'Image',
    accessor: (props) => props.image,
    disableSortBy: true,
  },
  {
    Header: 'Description',
    accessor: (props) => props.description,
    disableSortBy: true,
  },
];
