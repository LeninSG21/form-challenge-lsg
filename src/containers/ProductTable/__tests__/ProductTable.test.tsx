import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ProductsApi } from '../../../api/ProductsApi';
import { ProductsModel } from '../../../data-types/models';
import { ProductTable } from '../ProductTable';

const renderComponent = () => <ProductTable />;

const mockProductsModel: ProductsModel = [
  {
    code: 'HWR16-03',
    position: 1,
    quantity: 1,
    image: '01.png',
    price: 500,
    description:
      'Upright for room designs incl. floor profile in stainless steel appearance, depth: 250 - 900 mm, edging on both sides, Plinth height 70 mm: 2253 mm Plinth height 100 mm: 2283 mm Plinth height 150 mm: 2333 mm Plinth height 200 mm: 2383 mm',
  },
];

const mockError = new Error('Mock error');

describe('ProductTable', () => {
  const spy = jest.spyOn(ProductsApi, 'getProducts');

  beforeEach(() => {
    spy.mockClear();
  });

  describe('loading', () => {
    it('should render loading message when fetching', async () => {
      // arrange
      jest.useFakeTimers();
      spy.mockImplementationOnce(
        () =>
          new Promise((res) => {
            setTimeout(() => res(mockProductsModel), 1000);
          }),
      );
      render(renderComponent());

      // assert
      expect(screen.getByText(/Loading/)).toBeInTheDocument();
      jest.advanceTimersToNextTimer();
      await waitFor(() =>
        expect(screen.queryByText(/Loading/)).not.toBeInTheDocument(),
      );
      jest.useRealTimers();
    });
  });

  describe('rejected', () => {
    it('should render retry button when rejected', async () => {
      // arrange
      spy.mockRejectedValueOnce(mockError);
      render(renderComponent());

      // assert
      expect(
        await screen.findByRole('button', { name: /Retry/ }),
      ).toBeInTheDocument();
    });

    it('should refetch when retry button is clicked', async () => {
      // arrange
      spy.mockRejectedValueOnce(mockError);
      render(renderComponent());

      // act
      const button = await screen.findByRole('button', { name: /Retry/ });
      spy.mockResolvedValue(mockProductsModel);
      userEvent.click(button);

      // assert
      await waitFor(() => expect(spy).toHaveBeenCalledTimes(2));
    });
  });

  describe('Table', () => {
    beforeEach(() => {
      spy.mockResolvedValue(mockProductsModel);
    });

    it('should match table snapshot', async () => {
      // arrange
      const { container } = render(renderComponent());

      // act
      await waitFor(() =>
        expect(screen.queryByText(/Loading/)).not.toBeInTheDocument(),
      );

      // assert
      expect(container).toMatchSnapshot();
    });
  });
});
