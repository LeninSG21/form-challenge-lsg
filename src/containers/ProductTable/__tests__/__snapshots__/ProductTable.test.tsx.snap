// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTable Table should match table snapshot 1`] = `
<div>
  <div
    class="general-filters"
  >
    <input
      placeholder="Search 1 items..."
      value=""
    />
  </div>
  <table
    role="table"
    style="min-width: 0px;"
  >
    <thead>
      <tr
        role="row"
        style="display: flex; flex: 1 0 auto; min-width: 0px;"
      >
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          <div
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Code
             ↕️
          </div>
          <div>
            <input
              class="substring-filter"
              placeholder="Search 1 items..."
              value=""
            />
          </div>
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          <div
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Position
             ↕️
          </div>
          <div>
            <div
              class="range-filter"
            >
              <input
                placeholder="Min"
                type="number"
              />
              to
              <input
                placeholder="Max"
                type="number"
              />
            </div>
          </div>
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          <div
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Quantity
             ↕️
          </div>
          <div>
            <div
              class="range-filter"
            >
              <input
                placeholder="Min"
                type="number"
              />
              to
              <input
                placeholder="Max"
                type="number"
              />
            </div>
          </div>
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          <div
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Price
             ↕️
          </div>
          <div>
            <div
              class="range-filter"
            >
              <input
                placeholder="Min"
                type="number"
              />
              to
              <input
                placeholder="Max"
                type="number"
              />
            </div>
          </div>
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 300 0 auto; min-width: 0px; width: 300px;"
        >
          <div>
            Image
          </div>
          <div />
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 300 0 auto; min-width: 0px; width: 300px;"
        >
          <div>
            Description
          </div>
          <div />
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
        style="display: flex; flex: 1 0 auto; min-width: 0px;"
      >
        <td
          role="cell"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          HWR16-03
        </td>
        <td
          role="cell"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          1
        </td>
        <td
          role="cell"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          1
        </td>
        <td
          role="cell"
          style="box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px;"
        >
          500
        </td>
        <td
          role="cell"
          style="box-sizing: border-box; flex: 300 0 auto; min-width: 0px; width: 300px;"
        >
          <img
            alt="01.png"
            src="01.png"
          />
        </td>
        <td
          role="cell"
          style="box-sizing: border-box; flex: 300 0 auto; min-width: 0px; width: 300px;"
        >
          Upright for room designs incl. floor profile in stainless steel appearance, depth: 250 - 900 mm, edging on both sides, Plinth height 70 mm: 2253 mm Plinth height 100 mm: 2283 mm Plinth height 150 mm: 2333 mm Plinth height 200 mm: 2383 mm
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
